### PR TITLE
[ADD] New module sale_order_credit_limit_approval

### PR DESCRIPTION
**Pull request by forvas**
This pull request adds the module `sale_order_credit_limit_approval`.

This module modifies `res_partner` model and form view, in addition to `sale_order` model, form view and workflow.
- It creates a new state in sales workflow, named `waiting_for_approval`. When creating a sale order, if the credit limit of the customer is exceeded, a message will be shown in the form. When confirming this sale order, it will be passed to `waiting for approval state`. In this state, only sales managers will be able to approve or deny the order and continue the standard workflow.
- It also makes credit limit optional, and shows the available credit of the customers in the partners form.

More info in the description of the module at /static/description/index.html.

### CODE REVIEW

File: credit_limit_sale_order_approval/i18n/credit_limit_sale_order_approval.pot [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Translation of Odoo Server.
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/i18n/credit_limit_sale_order_approval.pot]: There's no need now of the pot file with Transifex process. Remove it for clarity.

File: credit_limit_sale_order_approval/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#    __openerp__.py file at the root folder of this module.    #
+################################################################
+
+from openerp import models, fields, api
+import logging
+_logger = logging.getLogger(__name__)
+
+
+# RES.PARTNER
+class res_partner(models.Model):
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/res_partner.py]: Use CamelCase for class names

File: credit_limit_sale_order_approval/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @api.multi
+    @api.depends('credit', 'credit_limit')
+    def _get_available_credit(self):
+        for partner in self:
+            available_credit = partner.credit_limit - partner.credit
+            if partner.set_credit_limit and available_credit > 0:
+                partner.available_credit = available_credit
+            else:
+                partner.available_credit = 0
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/res_partner.py]: All computed fields are reset at the beginning of the calculation to their default, so there's no need of putting this line.

File: credit_limit_sale_order_approval/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+_logger = logging.getLogger(__name__)
+
+
+# RES.PARTNER
+class res_partner(models.Model):
+    _inherit = 'res.partner'
+
+    @api.multi
+    @api.depends('credit', 'credit_limit')
+    def _get_available_credit(self):
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/res_partner.py]: Name the compute methods with `_compute_field_name` convention

File: credit_limit_sale_order_approval/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
+################################################################
+#    License, author and contributors information in:          #
+#    __openerp__.py file at the root folder of this module.    #
+################################################################
+
+from openerp import models, fields, api
+import logging
+_logger = logging.getLogger(__name__)
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/res_partner.py]: Don't add this by default if it's not used.

File: credit_limit_sale_order_approval/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+################################################################
+#    License, author and contributors information in:          #
+#    __openerp__.py file at the root folder of this module.    #
+################################################################
+
+from openerp import models, fields, api, _
+import logging
+_logger = logging.getLogger(__name__)
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/sale.py]: Don't add this by default if it's not used.

File: credit_limit_sale_order_approval/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+# RES.PARTNER
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/res_partner.py]: This comment is adding no value

File: credit_limit_sale_order_approval/models/sale.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+
+# SALE.ORDER
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/sale.py]: This comment is adding no value

File: credit_limit_sale_order_approval/models/sale.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+class sale_order(models.Model):
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/sale.py]: CamelCase

File: credit_limit_sale_order_approval/models/sale.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            else:
+                sale_order.credit_limit_warning = _(
+                    'Contact credit limit exceeded.')
+
+    confirmation_allowed = fields.Boolean(string='Confirmation allowed',
+                                          compute=_get_confirmation_allowed)
+    credit_limit_warning = fields.Char(string='Warning', compute=_get_warning)
+
+    state = fields.Selection(
+        selection=[('draft', 'Draft Quotation'),
```
Comment (pedrobaeza) [inline credit_limit_sale_order_approval/models/sale.py]: Don't overwrite full field, but use add_selection key for adding a value to the list.

**Comment by pedrobaeza** [comment]
Hi, Juan, thanks for your first contribution to OCA! I'm glad to see doing the next step.

I have put some comments to be followed across the module. Please take a look.

Also a proper name for the module should be _sale_order_credit_limit_approval_. We did something similar with higher scope, but it's a WIP, in https://github.com/odoomrp/odoomrp-wip/pull/152. Did you check?

**Comment by forvas** [comment]
Hello @pedrobaeza : 

Thank you for advices, I have corrected all issues in my new pull request https://github.com/OCA/sale-workflow/commit/2ca1a2b24e283809b78ebb7a9f2fa3e8a3e67934.

I have not checked the module you mentioned, do you think this one I made is going to be redundant?

**Comment by pedrobaeza** [comment]
Well, that one also cover the sale order approval, but if we split the process in several modules, there's no special problem. There's another interesting concept in that module that is the commercial risk, that takes into account also confirmed sales order amount that you can check.

**Comment by andhit-r** [comment]
@pedrobaeza is https://github.com/odoomrp/odoomrp-wip/pull/152 same as https://github.com/OCA/partner-contact/tree/9.0/partner_financial_risk? Is there any blueprint regarding credit limit that i can follow and help?

**Comment by pedrobaeza** [comment]
@andhit-r, both serves the same functionality, but we have evolved the concept. No blueprint unfortunately.

**Comment by andhit-r** [comment]
@forvas @pedrobaeza i backported partner_financial_risk to 8.0 (https://github.com/OCA/partner-contact/pull/361). After that PR merge, i will backport partner_sale_risk and partner_stock_risk. Reviews are most welcome.

**Comment by pedrobaeza** [comment]
Merged, so this one can be closed.

**Pull request closed without merging**